src/main/java/com/allenai/ml/linalg/Vector.java

```diff
@@ -0,0 +1,238 @@
+            result[i] = at(dimensionIndices[i]);
+        }
+        return result;
+    }
+
+    /**
+     * Mutate the Vector-in-place according to `updateFn`
+     * @param updateFn
+     */
+    default void mapInPlace(EntryUpdateFunction updateFn) {
+        entries().forEachOrdered(entry -> {
+            double updatedValue = updateFn.update(entry.index, entry.value);
+            set(entry.index, updatedValue);
+        });
+    }
+
+    /**
+     *
+     * @param updateFn
+     * @return Copy of the Vector with each entry having `updateFn` applied
+     */
+    default Vector map(EntryUpdateFunction updateFn) {
+        Vector result = copy();
+        result.mapInPlace(updateFn);
+        return result;
+    }
+
+    /**
+     * In place update of `scale  * x_i + offset` for each element.
+     * @param scale
+     * @param offset
+     */
+    default void affineUpdateInPlace(double scale, double offset) {
+        this.mapInPlace((idx, value) -> scale * value + offset);
+    }
+
+    /**
+     * Like [`affineUpdateInPlace`] but
+     * @param scale
+     * @param offset
+     * @return
+     */
+    default Vector affine(double scale, double offset) {
+      Vector result = copy();
+      result.affineUpdateInPlace(scale, offset);
+      return result;
+    }
+
+    /**
+     * @param scale
+     * @return Copy of the vector scaled by `scale`
+     */
+    default Vector scale(double scale) {
+        return affine(scale, 0.0);
+    }
+
+    /**
+    * Compute the [dot-product](http://en.wikipedia.org/wiki/Dot_product) with `other` vector
+    * @param other
+    * @return
+    */
+    default double dotProduct(Vector other) {
+      if (other.dimension() != this.dimension()) {
+          throw new IllegalArgumentException("Dimensions don't match");
+      }
+      // Do dot-product from the pov of vector with fewer elements
+      if (other.numStoredEntries() < this.numStoredEntries()) {
+        return other.dotProduct(this);
+      }
+      return nonZeroEntries()
+        .mapToDouble(entry -> entry.value * other.at(entry.index))
+        .sum();
+    }
+
+    default double l2NormSquared() {
+        return this.dotProduct(this);
+    }
+
+    default double l2Distance(Vector other) {
+        return this.add(-1.0, other).l2NormSquared();
+    }
+
+    default boolean closeTo(Vector other, double thresh) {
+        return this.l2Distance(other) < thresh;
+    }
+
+    double DIST_THRESH = 1.0e-10;
+
+    default boolean closeTo(Vector other) {
+        return closeTo(other, DIST_THRESH);
+    }
+
+
+    default double inc(long idx, double amount) {
+        double newValue = at(idx) + amount;
+        set(idx, newValue);
+        return newValue;
+    }
+
+    default Stream<Entry> entries() {
+        val spliterator = new VectorSpliterator(this, 0, dimension());
+        return StreamSupport.stream(spliterator, true);
+    }
+
+    default Stream<Entry> nonZeroEntries() {
+        return entries().filter(e -> e.value != 0.0);
```

Comment:
Mathematically, this is only correct for 0.0; it's just a space optimization to prune the unnecessary entries. For instance, if we truncated below `Math.abs(x) < 1.0e-10` then doing `vec.scale(1.0e10)` would yield the wrong answer. 

Note there isn't a notion of equality here, we use `close()` as a boolean check for a vector is close in L2 distance.